contrail-controller/hooks/contrail_controller_utils.py

```diff
@@ -179,3 +185,66 @@ def update_charm_status():
         docker_utils.compose_run(REDIS_CONFIGS_PATH + "/docker-compose.yaml")
 
     common_utils.update_services_status(SERVICES)
+
+
+def update_hosts_file(hosts_map):
+    """Update /etc/hosts file with cluster names and IPs.
+
+    RabbitMQ requires NODE names in a cluster to be resolvable.
+    https://www.rabbitmq.com/clustering.html#issues-hostname
+
+    In a multi-homed host scenario cluster IPs may have FQDNs structured
+    as interface_name.host_name.domain which will result in an issue if
+    a short name is derived from an FQDN by taking its first part.
+    See https://github.com/Juniper/contrail-charms/issues/50
+
+    This function updates /etc/hosts file with resolutions for IP -> hostname
+    lookups.
+    """
+    with open(HOSTS_FILE, 'r') as hosts:
+        lines = hosts.readlines()
+
+    log("Updating hosts file with: %s (current: %s)" % (hosts_map, lines),
+        level=INFO)
+
+    newlines = []
+    for ip, hostname in hosts_map.items():
+        if not ip or not hostname:
+            continue
+
+        keepers = []
+        for line in lines:
+            _line = line.split()
+            if len(line) < 2 or not (_line[0] == ip or hostname in _line[1:]):
+                keepers.append(line)
+            else:
+                log("Removing line '%s' from hosts file" % (line))
+
+        lines = keepers
+        newlines.append("%s %s\n" % (ip, hostname))
+
+    lines += newlines
+
+    with tempfile.NamedTemporaryFile(delete=False) as tmpfile:
+        with open(tmpfile.name, 'w') as hosts:
+            for line in lines:
+                hosts.write(line)
+
+    os.rename(tmpfile.name, HOSTS_FILE)
```

Comment:
why did you make writing with renaming?
if this operation should be atomic - then it should be done in the same folder. rename is not atomic across different disks that can be in case /tmp and /etc.

Reply:
Just reused the code from rabbitmq-charm.

The original commit isn't very descriptive but I would assume that the idea is to make the replacement atomic to avoid the situation where application runtimes (glibc etc.) would be reading the hosts file while it is being modified by the charm process. 

https://github.com/openstack/charm-rabbitmq-server/commit/8272f214ab2190c17cf8c84f7cd9f227dcabdf47#diff-21870f9ed3cd89ae3ca8d1f237afc315R384

I can pass the dir argument to NamedTemporaryFile to cover the case you described.